### PR TITLE
[Blocked] Update TransactionRequest's `to` field to TxKind

### PR DESCRIPTION
**Pull request by EmperorOrokuSaki**
## Motivation
Issue https://github.com/alloy-rs/alloy/issues/549
## Solution
Note: This is not the full refactor, I'll mark it ready for review once it's done :)
From the issue:

[X]  Modify to field to make creates explicit
[X]  Error on non-4844 build if to is unset
[X]  Error on 4844 build if to is unset or create
[X] Add following to the builder trait
```
    - [X]  fn as_create(self) -> Self shortcut to set the field to Some(Create)

    - [X] fn deploy_code(self, code: Vec<u8>) -> Self to set data and to fields at once

    - [X] fn with_call<T: SolCall>(self, t: T) -> Self to set data and ensure to is NOT Create (set to to some if it is create)
```

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ X] Breaking changes

### CODE REVIEW

File: crates/rpc-types/src/eth/transaction/request.rs [hunk 1]
```diff
@@ -460,6 +460,18 @@ impl From<TxEnvelope> for TransactionRequest {
...
+#[error("the field `to` is unset. Please populate it.")]
+#[non_exhaustive]
+pub struct TransactionRequestNon4844Error;
+
+/// Error thrown when `to` is `None` in non-EIP4844 transaction requests 
+#[derive(Debug, Default, thiserror::Error)]
+#[error("the field `to` can only be of type TxKind::Call(Account). Please change it accordingly.")]
+#[non_exhaustive]
+pub struct TransactionRequest4844Error;
+
```
Comment (prestwich) [inline crates/rpc-types/src/eth/transaction/request.rs:474]: so far we have not introduced explicit error types for most tx request fields, we've been using strings for missing keys. So i'd be fine if these structs were omitted

File: crates/rpc-types/src/eth/transaction/request.rs [hunk 2]
```diff
@@ -211,12 +212,14 @@ impl TransactionRequest {
     /// If required fields are missing. Use `complete_legacy` to check if the
     /// request can be built.
     fn build_legacy(self) -> TxLegacy {
+        let checked_to = self.to.expect("the `to` field should have value.");
```
Comment (prestwich) [inline crates/rpc-types/src/eth/transaction/request.rs:215]: let's make these match the others `to: self.to.expect("checked in complete_legacy"),`

**Review by prestwich — CHANGES_REQUESTED** [review]
looking good overall. need to add a `to` check to `fn check_reqd_fields` as well

**Comment by prestwich** [comment]
mututally dependent with https://github.com/alloy-rs/core/pull/606

File: crates/contract/src/call.rs [hunk 3]
```diff
@@ -252,12 +256,33 @@ impl<T: Transport + Clone, P: Provider<T, N>, N: Network> RawCallBuilder<T, P, N
...
+        Self::new_inner_call(provider, input, ())
+    }
+
+    /// Creates a new call builder with the provided provider and contract
+    /// deploy code.
+    ///
+    /// Will not decode the output of the call, meaning that
+    /// [`call`](Self::call) will behave the same as
+    /// [`call_raw`](Self::call_raw).
+    pub fn new_raw_deploy(provider: P, input: Bytes) -> Self {
```
Comment (DaniPopes) [inline crates/contract/src/call.rs:268]: Can you add the sol! NOTE comment (see new_sol) ?

File: crates/provider/src/fillers/gas.rs [hunk 4]
```diff
@@ -266,7 +266,7 @@ mod tests {
         let tx = TransactionRequest {
             from: Some(anvil.addresses()[0]),
             value: Some(U256::from(100)),
-            to: address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into(),
+            to: Some(address!("d8dA6BF26964aF9D7eEd9e03E53415D37aA96045").into()),
```
Comment (gakonst) [inline crates/provider/src/fillers/gas.rs:269]: is the `.into()` required?

Comment (prestwich) [reply crates/provider/src/fillers/gas.rs:269]: yeah but its just tests so i didn't want to clutter the diff

File: crates/contract/src/call.rs (hunk 3 above)
Comment (prestwich) [reply crates/contract/src/call.rs:268]: 🖖

**Comment by prestwich** [comment]
currently blocked by updating alloy-core to the patched version

**Comment by gakonst** [comment]
Unblocked now https://github.com/alloy-rs/core/commit/9ff1e5a58be01f4846d839f558b08366f8ce7c0f

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**